d3-series/materials/Week 4/Assignment/js/air_quality_chart.js

```diff
@@ -0,0 +1,147 @@
+
+        // scales
+        var x = d3
+            .scaleBand()
+            .domain(
+                data.map(function(d) {
+                    return d.State;
+                })
+            )
+            .range([0, innerWidth])
+            .padding(0.4);
+
+        var y = d3
+            .scaleLinear()
+            .domain([
+                0,
+                d3.max(data, function(d) {
+                    return d.Emissions*1.05;
+                })
+            ])
+            .range([innerHeight, 0]);
+
+        // axes
+        var xAxis = d3.axisBottom(x)
+
+        g
+            .append('g')
+            .attr('class', 'x-axis')
+            .attr('transform', 'translate(0,' + innerHeight + ')')
+            .call(xAxis);
+
+        var yAxis = d3.axisLeft(y).ticks(10);
+
+        g
+            .append('g')
+            .attr('class', 'y-axis')
+            .call(yAxis);
+
+        // bars
+        var colors = d3.scaleOrdinal()
+            .domain(["West", "South", "Midwest", "Northeast"])
+            .range(["#E69F00", "#009E73", "#56B4E9",  "#0072B2"]);
+
+        g
+            .selectAll('.bar')
+            .data(data)
+            .enter()
+            .append('rect')
+            .attr('class', 'bar')
+            .attr("fill", function(d) {
+                return colors(d.Region);
+            })
+            .attr('x', function(d) {
+                return x(d.State);
+            })
+            .attr('y', function(d) {
+                return y(d.Emissions);
+            })
+            .attr('width', x.bandwidth())
+            .attr('height', function(d) {
+                return innerHeight - y(d.Emissions);
+            })
+            .attr('stroke', 'none');
+
+
+        var createHorizontalLine = function(y_pos, line_length){
+            g
+                .append("g")
+                .attr("transform", "translate(0, "+ y(y_pos) +")")
+                .append("line")
+                .attr("x2", line_length)
+                .style("stroke", "black")
+                .style("stroke-width", "1px")
+        };
+
+        createHorizontalLine(0, innerWidth);
+
+        // axis labels
+        g
+            .append('text')
+            .attr('class', 'y-axis-label')
+            .attr('x', -30)
+            .attr('y', innerHeight / 2 - 40)
+            .attr('transform', 'rotate(-90,-30,' + innerHeight / 2 + ')')
+            .attr('text-anchor', 'middle')
+            .attr('dominant-baseline', 'baseline')
+            .style('font-size', 14)
+            .text('Million metric tons of energy-related carbon dioxide');
+
+        // title
+        g
+            .append('text')
+            .attr('class', 'title')
+            .attr('x', innerWidth / 2)
+            .attr('y', -20)
+            .attr('text-anchor', 'middle')
+            .attr('dominant-baseline', 'baseline')
+            .style('font-size', 20)
+            .text('2019 Annual Emissions');
+    });
+}
+
+buildChart('#air-quality-chart');
```

Comment:
Couple of ideas if time for enhancements/what would be expected in production code:
- break down the element generation in their own methods
- give elements classes and ids and style them with css stylesheet if the styling is not dynamic (for example text elements, axis
- avoid 'magic numbers' in middle of the code, declare them at the beginning of the method, class. Makes thing more manageable when those numbers need to change (don't have to chase them in the code).